README.md

```diff
@@ -1,2 +1,307 @@
+```
+
+
+## Input files
+
+The input files for Woltka are sequence **alignment** files. The term "alignment" here describes the operation of aligning short sequencing reads against long reference sequences. The basic information an alignment file provides is the **mapping** between queries and subjects. In addition, the position and quality of alignments are available in some formats, which are useful in some applications.
+
+### File formats
+
+Woltka supports the following alignment formats (specified by parameter `--format` or `-f`):
+
+- `map`: Simple mapping of query \<tab\> subject.
+- `sam`: SAM format. Supported by multiple tools such as Bowtie2 and BWA.
+- `b6o`: BLAST tabular format (i.e., BLAST parameter `-outfmt 6`). Supported by multiple tools such as BLAST, DIAMOND, VSEARCH, BURST, etc.
+
+If not specified, Woltka will automatically infer the format of input alignment files.
+
+Woltka supports and automatically detects common file compression formats including `gzip`, `bzip2` and `xz`.
+
+### Filename patterns
+
+In the default mode, Woltka treats every file under the directory specified by the `--input` or `-i` parameter as an alignment file for one sample. The sample ID is automatically extracted from the filename, with filename extension stripped. Compression file extensions are automatically recognized. For example, the sample IDs for `S01.sam` and `S02.m8.gz` are `S01` and `S02`, respectively.
+
+One may restrict this behavior to avoid confusions (e.g., there are unwanted files) by the following two parameters:
+
+- `--filext` or `-e`: A suffix to be stripped from filenames, and the remaining part is considered a sample ID.
+
+  For example, if valid alignment filenames have the pattern of `ID_L001.aln.bz2`, one may specify `-e _L001.aln.bz2`, so that only `ID` is retained, and filenames which do not have this suffix (e.g., `ID.bt2.log` or`readme.txt`) are ignored.
+
+- `--sample-ids` or `-s`: A file containing valid sample IDs (one ID per line) to include in the current analysis. This list also specifies the order of sample IDs in the output profile.
+
+  Only the first column before \<tab\> is considered. Lines starting with `#` are omitted. Therefore, a metadata table may also serve as a valid sample ID list.
+
+### Demultiplexing
+
+Woltka supports convenient demultiplexing. If the input path (specified by `--input` or `-i`) points to a file instead of a directory, Woltka will treat it as a multiplexed alignment file.
+
+Specifically, the program divides sample ID and read ID by the first underscore in each query identifier in the alignment file (i.e., a pattern of `sample_read`).
+
+One may manually switch on or off the demultiplexing function by adding flag `--demux` or `--no-demux` to the command. This is useful when there are several multiplexed alignment files (e.g., each from one sequencing lane) in one **directory**, or the only input alignment **file** provided is not multiplexed but just for a single sample.
```

Comment:
It may be better to have a single default behavior, as opposed to changing default based on other inputs. Goes in the same vein as "better to require more user input than to infer wrong".

Reply:
I had some relevant discussion above to the "sample ID - aligment file" question, which also applies here. I guess a good way to balance between "smart" and "strict" is to have the program print out auto-determined parameter status.

Meanwhile, I admit that the current design is bit too complicated. Perhaps a better way is to have more than one `input` parameter: `--input-dir` is for multiple samples, `--input-file` is for one multiplexed file, `--input-map` is for a sample-to-alignment mapping file. But I am not sure if this way is compatible with QIIME 2 (it seems to me that all input files are mandatory).

Also I feel that `click`'s design of boolean flags is quite interesting: it actually allows three values: `True`, `False` and `None`, and it is so for a reason...

Reply:
I think qiime2 should be able to support the different input requirements with `TypeMap`

See this thread for a quick demo of `TypeMap`: https://github.com/qiime2/q2-quality-control/pull/47

If you were to implement the `--input-dir` and `--input-file` feature (which I think is a good idea), there might be a need to have multiple Q2 commands (one for multiplexed file, one for demultiplexed files), but I think that may be fine... Note that you can specify that parameters are optional with Qiime2 (it infers this from the type hints in your signatures). But it might be a bit awkward since both `input-dir` and `input-map` could be considered as optional, but you need at least one...
Q2 use could also be restricted to giving demultiplexed files since this can be done through qiime2, and it is done in a way that will track the provenance, which is a major bonus of using q2.

Reply:
@gwarmstrong Thanks for pointing me to `TypeMap`! Will explore that!

A challenge in (de)multiplexing is that alignments are usually huuuuge. It is ideal to leave them as is rather than reformatting before one can run Woltka.

Btw, I don't know how `--input-dir` works in QIIME 2. Is it not an input (i.e., `--i-`) but a parameter (`--p-`)? For the reason mentioned above, the input file(s) for Woltka are likely not Artifacts but raw files.

Reply:
That is true... I think that the current structure is very incompatible with how we usually think of qiime plugins.... I think we could apply some careful thought as to how best make a plugin that achieves the reproducibility aspect of qiime2. Since the operation of the plugin right now kind of breaks that...